contracts/validators/request/CredentialAtomicQueryV3Validator.sol

```diff
@@ -0,0 +1,326 @@
+
+        IGroth16Verifier g16Verifier = getVerifierByCircuitId(credAtomicQuery.circuitIds[0]);
+        if (g16Verifier == IGroth16Verifier(address(0))) {
+            revert VerifierAddressShouldNotBeZero();
+        }
+
+        // verify that zkp is valid
+        if (!g16Verifier.verify(a, b, c, inputs)) {
+            revert ProofIsNotValid();
+        }
+
+        return parsePubSignals(inputs);
+    }
+
+    function _checkPubSignalsWithCredentialAtomicQuery(
+        CredentialAtomicQueryV3 memory credAtomicQuery,
+        PubSignals memory pubSignals,
+        address sender
+    ) internal view returns (IRequestValidator.ResponseField[] memory) {
+        _checkAllowedIssuers(pubSignals.issuerID, credAtomicQuery.allowedIssuers);
+        _checkProofExpiration(pubSignals.timestamp);
+
+        _checkGroupIDOrLinkID(credAtomicQuery.groupID, pubSignals.linkID);
+        _checkProofType(credAtomicQuery.proofType, pubSignals.proofType);
+        _checkNullify(pubSignals.nullifier, credAtomicQuery.nullifierSessionID);
+
+        // GIST root and state checks
+        _checkClaimIssuanceState(pubSignals.issuerID, pubSignals.issuerState);
+        _checkClaimNonRevState(pubSignals.issuerID, pubSignals.issuerClaimNonRevState);
+        if (pubSignals.isBJJAuthEnabled == 1) {
+            _checkGistRoot(pubSignals.userID, pubSignals.gistRoot);
+        } else {
+            _checkAuth(pubSignals.userID, sender);
+        }
+
+        // Checking challenge to prevent replay attacks from other addresses
+        _checkChallenge(pubSignals.challenge, sender);
+
+        // check circuitQueryHash
+        if (pubSignals.circuitQueryHash != credAtomicQuery.queryHash) {
+            revert QueryHashDoesNotMatchTheRequestedOne(
+                credAtomicQuery.queryHash,
+                pubSignals.circuitQueryHash
+            );
+        }
+
+        // if operator == 16 then we have selective disclosure
+        return _getResponseFields(pubSignals, credAtomicQuery.operator == 16);
+    }
+
+    function _checkGroupIDOrLinkID(uint256 groupID, uint256 linkID) internal pure {
+        if (groupID == 0 && linkID == 0) return;
+        if (groupID != 0 && linkID != 0) return;
+        revert InvalidGroupIDOrLinkID(groupID, linkID);
+    }
+
+    function _checkProofType(uint256 queryProofType, uint256 pubSignalProofType) internal pure {
+        if (queryProofType != 0 && queryProofType != pubSignalProofType) {
+            revert ProofTypeShouldMatchTheRequestedOneInQuery();
+        }
+    }
+
+    function _checkNullify(uint256 nullifier, uint256 nullifierSessionID) internal pure {
+        if (nullifierSessionID != 0 && nullifier == 0) {
+            revert InvalidNullifyPubSignal();
+        }
+    }
+
+    function _checkAuth(uint256 userID, address ethIdentityOwner) internal view {
+        if (
+            userID !=
+            GenesisUtils.calcIdFromEthAddress(
+                _getState().getIdTypeIfSupported(userID),
```

Comment:
Probably doesn't work for cross-chain proofs.

Reply:
It's a good point. Better get `idType` via `GenesisUtils.getIdType(..)` as verification chain may not support user source chain `idType`